### PR TITLE
updating of functions to work as rows-last

### LINKED ISSUE #364: addRow function appending to wrong dimension of multidimensional matrices
Given MATLAB's rows-last mode of input-output with NWB files, the .addRow method of `DynamicTable` objects should append to last dimension of multi-dimensional columns. 

An example:

```matlab
file= NwbFile( ...
    'session_start_time', '2021-01-01 00:00:00', ...
    'identifier', 'ident1', ...
    'session_description', 'test' ...
    );

colnames = {'start_time', 'stop_time', 'randomvalues'};
multi_dim_table1 = types.core.TimeIntervals(...
    'description', 'test dynamic table column',...
    'colnames', colnames);
            
for i = 1:100
    multi_dim_table1.addRow(...
        'start_time', i,...
        'stop_time', i+1,...
        'randomvalues',rand(1,3,2),...
        'id', i-1);
end
file.intervals.set('multi', multi_dim_table1);
nwbExport(file, 'test_multi_addRow.nwb');
```
The code above produces a 100x3x2 array for the 'randomvalues' column which is encoded into file as 3x2x100 (see below). 

Screen Shot 2021-12-14 at 11 52 31 AM

The correct behavior would append each 3x2 matrix to the last dimension to produce a 3x2x100 array that is encoded into file as a 100x3x2 array.

This modification would resolve the concerns brought up in the discussion of issue #360

### PR DESCRIPTION
**Pull request by cechava**
fixes #364 

The following should be amended to work with a rows-last convention:

- [x] checkConfig
- [x] getRow
- [x] addRow

make sure to have tests for DynamicTable columns of following varieties:
- [x] one-dimensional columns
- [x] multi-dimensional columns
- [x] ragged multi-dimensional columns
- [x] nested ragged columns
- [x] expandable columns (DataPipe)

### CODE REVIEW

**Comment by cechava** [comment]
with this first commit, valid Dynamic Tables can be defined in the following manner: 

```matlab
>> col1 = types.hdmf_common.VectorData( ...
    'description', 'column #1', ...
    'data', [1, 2] ...
);

col2 = types.hdmf_common.VectorData( ...
    'description', 'column #2', ...
    'data', {'a', 'b'} ...
);

my_table = types.hdmf_common.DynamicTable( ...
    'description','an example table', ...
    'colnames', {'col1','col2'}, ...
    'col1',col1, ...
    'col2',col2, ...
    'id', types.hdmf_common.ElementIdentifiers('data', [0, 1]) ...  % 0-indexed, for compatibility with Python
);
```

getRow needs to be updated, since current output is as follows:

```matlab
>> my_table.getRow(1)

ans =

  1×2 table

     col1          col2     
    ______    ______________

    1    2    {'a'}    {'b'}
```

**Comment by cechava** [comment]
with latest commit, getRow has correct behavior for rows-last convention (at least for simple tables):

```matlab
>> my_table.getRow(2)

ans =

  1×2 table

    col1    col2 
    ____    _____

     2      {'b'}

>> my_table.getRow(1)

ans =

  1×2 table

    col1    col2 
    ____    _____

     1      {'a'}
```

**Comment by cechava** [comment]
with latest commit, the last dimension of `DynamicTable` objets are printed along the vertical axes (i.e., as conventional rows)

```matlab

start_times = types.hdmf_common.VectorData( ...
    'description', 'start_times', ...
    'data', 1:10 ...
);

stop_times = types.hdmf_common.VectorData( ...
    'description', 'stop_times', ...
    'data', 2:11 ...
);
randomvalues = types.hdmf_common.VectorData( ...
    'description', 'random values',...
    'data', rand(3,2,10) ...
);

multi_dim_table = types.core.TimeIntervals( ...
    'description','test table', ...
    'colnames', {'randomvalues', 'start_time', 'stop_time'}, ...
    'start_time', start_times, ...
    'stop_time', stop_times, ...
    'randomvalues', randomvalues, ...
     'id', types.hdmf_common.ElementIdentifiers('data', 0:9) ...
 );

>> multi_dim_table.toTable()

ans =

  10×4 table

    id     randomvalues     start_time    stop_time
    __    ______________    __________    _________

    0     [1×2×3 double]         1            2    
    1     [1×2×3 double]         2            3    
    2     [1×2×3 double]         3            4    
    3     [1×2×3 double]         4            5    
    4     [1×2×3 double]         5            6    
    5     [1×2×3 double]         6            7    
    6     [1×2×3 double]         7            8    
    7     [1×2×3 double]         8            9    
    8     [1×2×3 double]         9           10    
    9     [1×2×3 double]        10           11    
```

**Comment by cechava** [comment]
with latest commit, can use the addRow method to add elements to the last dimension of a DynamicTable column
```matlab
col1 = types.hdmf_common.VectorData( ...
    'description', 'column #1', ...
    'data', [1, 2] ...
);

col2 = types.hdmf_common.VectorData( ...
    'description', 'column #2', ...
    'data', {'a', 'b'} ...
);

my_table = types.hdmf_common.DynamicTable( ...
    'description','an example table', ...
    'colnames', {'col1','col2'}, ...
    'col1',col1, ...
    'col2',col2, ...
    'id', types.hdmf_common.ElementIdentifiers('data', [0, 1]) ...  % 0-indexed, for compatibility with Python
);
my_table.addRow('col1', 3, 'col2', {'c'}, 'id', 2);

>> my_table.toTable()

ans =

  3×3 table

    id    col1    col2 
    __    ____    _____

    0      1      {'a'}
    1      2      {'b'}
    2      3      {'c'}

```
also works for ragged arrays:
```matlab
col1 = types.hdmf_common.VectorData( ...
    'description', 'column #1', ...
    'data', {'1a', '1b', '1c', '2a'} ...
);

col1_index = types.hdmf_common.VectorIndex( ...
    'description', 'column #1 index',...
    'target',types.untyped.ObjectView(col1), ...  % object view of target column
    'data', [3, 4] ...
);
% 
table_ragged_col = types.hdmf_common.DynamicTable( ...
    'description', 'an example table', ...
    'colnames', {'col1'}, ...
    'col1', col1, ...
    'col1_index', col1_index, ...
    'id', types.hdmf_common.ElementIdentifiers('data', [0, 1]) ...  % 0-indexed, for compatibility with Python
);

table_ragged_col.addRow('col1', {'3a','3b','3c'}, 'id', 2);

>> table_ragged_col.toTable()

ans =

  3×2 table

    id       col1   
    __    __________

    0     {1×3 cell}
    1     {1×1 cell}
    2     {1×3 cell}
```

**Comment by cechava** [comment]
with latest commit, addRow function works for vararg input and MATLAB table inputs:

```matlab
col1 = types.hdmf_common.VectorData( ...
    'description', 'column #1', ...
    'data', rand(3,2,10) ...
);

my_table = types.hdmf_common.DynamicTable( ...
    'description','an example table', ...
    'colnames', {'col1'}, ...
    'col1',col1, ...
    'id', types.hdmf_common.ElementIdentifiers('data', 0:9) ... 
);
% add row with vararg
my_table.addRow('col1',rand(3,2), 'id', 10);

% add row with table
col1 = rand(3,2,5);
sub_table = table(col1);
my_table.addRow(sub_table);

>> my_table.vectordata.get('col1')

ans = 

  VectorData with properties:

    description: 'column #1'
           data: [3×2×16 double]

>> my_table.toTable()

ans =

  14×2 table

    id         col1     
    __    ______________

     0    [1×2×3 double]
     1    [1×2×3 double]
     2    [1×2×3 double]
     3    [1×2×3 double]
     4    [1×2×3 double]
     5    [1×2×3 double]
     6    [1×2×3 double]
     7    [1×2×3 double]
     8    [1×2×3 double]
     9    [1×2×3 double]
    10    [1×2×3 double]
    11    [1×2×3 double]
    12    [1×2×3 double]
    13    [1×2×3 double]


```

**Comment by cechava** [comment]
@bendichter @ln-vidrio 

I've hit a bit of a wall in implementing changes to deal with a rows-last convention in MatNWB and I'm wondering if you have any thoughts. Let me illustrate the issue with an example. Let's say I'm trying to build a `DynamicTable` by progressively adding rows in the following manner: 

```matlab
colnames = {'start_time', 'stop_time', 'random_multi'};
time_table = types.core.TimeIntervals(...
    'description', 'test dynamic table column',...
    'colnames', colnames ...
);

for i = 1:10
    time_table.addRow( ...
        'start_time', i, ...
        'stop_time', i+1, ...
        'random_multi', rand(3,2), ...
        'id', i-1 ...
    )
end
```

After going through the first loop( and adding the first row) the checkConfig utility function will find an unmatched length of columns. 'start_time' and 'stop_time' will have length 1 while 'random_multi' will have length 2 because 1) there's only one 3x2 matrix in the 'random_multi' column and 2) MATLAB doesn't recognize singleton dimensions so that implicit 3x2x1 becomes 3x2. 

Right now, I'm dealing with this by looking at the other columns and making the decision to not assert equal length across all columns if there's a column of length one (in which case the dimensions of such multidimensional columns are ambiguous - is it 3x2 or 3x2x1). I don't foresee this solution being a problem, but let me know if you guys can think of a more elegant solution.

**Comment by cechava** [comment]
with latest commit, Dynmic Tables can be built up row-by-row 

```matlab

colnames = {'start_time', 'stop_time', 'random_multi'};
time_table = types.core.TimeIntervals(...
    'description', 'test dynamic table column',...
    'colnames', colnames ...
);

for i = 1:10
    time_table.addRow( ...
        'start_time', i, ...
        'stop_time', i+1, ...
        'random_multi', rand(3,2), ...
        'id', i-1 ...
    )
end

>> time_table.toTable()

ans =

  10×4 table

    id    start_time    stop_time     random_multi 
    __    __________    _________    ______________

    0          1            2        [1×2×3 double]
    1          2            3        [1×2×3 double]
    2          3            4        [1×2×3 double]
    3          4            5        [1×2×3 double]
    4          5            6        [1×2×3 double]
    5          6            7        [1×2×3 double]
    6          7            8        [1×2×3 double]
    7          8            9        [1×2×3 double]
    8          9           10        [1×2×3 double]
    9         10           11        [1×2×3 double]
```

**Comment by bendichter** [comment]
@cechava It's a shame MATLAB does not allow for singleton dimensions. That sounds like a good solution to me.

**Comment by cechava** [comment]
with latest commit, adding nested ragged arrays supported

```matlab

colnames = {'start_time', 'stop_time', 'randomvalues', 'random_multi','stringdata'};
time_table = types.core.TimeIntervals(...
    'description', 'test dynamic table column',...
    'colnames', colnames ...
);
% 
for i = 1:10
    time_table.addRow( ...
        'start_time', i, ...
        'stop_time', i+1, ...
        'random_multi', rand(3,2), ...
        'randomvalues', {rand(2,5), rand(2,3)}, ...
        'stringdata', {'TRUE'},...
        'id', i-1 ...
    )
end

>> time_table.vectordata.get('randomvalues')

ans = 

  VectorData with properties:

    description: 'AUTOGENERATED description for column `randomvalues`'
           data: [2×80 double]

>> time_table.toTable()

ans =

  10×6 table

    id    start_time    stop_time    randomvalues    random_multi    stringdata
    __    __________    _________    ____________    ____________    __________

    0          1            2         {1×2 cell}     {3×2 double}     {'TRUE'} 
    1          2            3         {1×2 cell}     {3×2 double}     {'TRUE'} 
    2          3            4         {1×2 cell}     {3×2 double}     {'TRUE'} 
    3          4            5         {1×2 cell}     {3×2 double}     {'TRUE'} 
    4          5            6         {1×2 cell}     {3×2 double}     {'TRUE'} 
    5          6            7         {1×2 cell}     {3×2 double}     {'TRUE'} 
    6          7            8         {1×2 cell}     {3×2 double}     {'TRUE'} 
    7          8            9         {1×2 cell}     {3×2 double}     {'TRUE'} 
    8          9           10         {1×2 cell}     {3×2 double}     {'TRUE'} 
    9         10           11         {1×2 cell}     {3×2 double}     {'TRUE'} 
```

**Comment by cechava** [comment]
@bendichter @ln-vidrio 

another potential complication is that we can't use MATLAB tables as arguments to addRows. This is because we can't create tables with an unmatched number of rows (first-dimension as MATLAB defines it). For example, the code below doesn't work anymore
```matlab
colnames = {'start_time', 'stop_time', 'randomvalues', 'random_multi','stringdata'};
time_table = types.core.TimeIntervals(...
    'description', 'test dynamic table column',...
    'colnames', colnames ...
);
subTable = table(...
    (101:150), ...
    (102:151), ...
    rand(2,3,50),...
    repmat({'TRUE'}, 1, 50),...
    'VariableNames', {'start_time', 'stop_time', 'random_multi','stringdata'});
time_table.addRow(subTable)

Error using table (line 233)
All table variables must have the same number of rows.
```

The issue is that the 'random_multi' column has 2 rows while all other columns have one row, as defined by MATLAB. One potential work-around is to have users define MATLAB tables with the number of rows along the first-dimension and under the hood we take that first dimension and map it onto the last dimension when using the addRow method.
I can't decide if this is a good idea because I think it will make things really confusing with users if they are using a rows-first convention for MATLAB tables and then using rows-last convention for matnwb tables. Perhaps the thing to do is to just remove the option of passing in MATLAB tables to the addRow method. Let me know your thoughts.

**Comment by bendichter** [comment]
@cechava Thanks for bringing this up. I did not think of this, and I see that it presents a problem. I think your idea of transposing under the hood is probably the best way we could handle this if we wanted to continue to support it, but I agree it is potentially confusing. I could go either way. I guess I would rather not allow tables as input, but I think it would be OK if you wanted to keep them. We'd need to be very clear about this in the documentation though

**Comment by cechava** [comment]
After sleeping on it, I agree with you and think it would be best to not allow tables as input. I can gently deprecate the addTableRow function by throwing out an error with a message whenever the function is called (e.g., adding rows by supplying a MATLAB table has been deprecated. Please use key-value pair arguments)

**Comment by cechava** [comment]
with latest commit, rows-last convention includes support for expandable tables
```matlab
% create NwbFile object with required fields
file= NwbFile( ...
    'session_start_time', '2021-01-01 00:00:00', ...
    'identifier', 'ident1', ...
    'session_description', 'ExpandableTableTutorial' ...
    );

% create VectorData objects with DataPipe objects
start_time_exp = types.hdmf_common.VectorData( ...
    'description', 'start times column', ...
    'data', types.untyped.DataPipe( ...
        'data', [1, 2], ...  # data must be numerical
        'maxSize', [1, Inf], ...
        'axis', 2 ...
    ) ...
);

stop_time_exp = types.hdmf_common.VectorData( ...
    'description', 'stop times column', ...
    'data', types.untyped.DataPipe( ...
        'data', [10, 20], ...  #data must be numerical
        'maxSize', [1, Inf], ...
        'axis', 2 ...
    ) ...
);

random_exp = types.hdmf_common.VectorData( ...
    'description', 'random data column', ...
    'data', types.untyped.DataPipe( ...
        'data', rand(5, 2), ...  #random data
        'maxSize', [5, Inf], ...
        'axis', 2 ...
    ) ...
);

ids_exp = types.hdmf_common.ElementIdentifiers( ...
    'data', types.untyped.DataPipe( ...
        'data', [0, 1], ...  # data must be numerical
        'maxSize', [1, Inf], ...
        'axis', 2 ...
    ) ...
);
%create expandable table
colnames = {'start_time', 'stop_time', 'randomvalues'};
file.intervals_trials = types.core.TimeIntervals( ...
    'description', 'test expdandable dynamic table', ...
    'colnames', colnames, ...
    'start_time', start_time_exp, ...
    'stop_time', stop_time_exp, ...
    'randomvalues', random_exp, ...
    'id', ids_exp ...
);  

% export file
nwbExport(file, 'expandableTableTestFile.nwb');
% % 
readFile = nwbRead('expandableTableTestFile.nwb');
readFile.intervals_trials.addRow( ...
    'start_time', 3, ...
    'stop_time', 30, ...
    'randomvalues', rand(5,1), ...
    'id', 2 ...
    )
nwbExport(readFile, 'expandableTableTestFile.nwb');
```
Screen Shot 2021-12-20 at 11 30 40 AM

**Comment by bendichter** [comment]
Great! Have you been updating the tutorial to reflect these changes?

**Comment by cechava** [comment]
yes, I've been updating a local copy of the tutorials to make sure that things function as expected with these changes. Once we settle and merge this PR, I will update the PR for the tutorials

**Comment by cechava** [comment]
@bendichter @ln-vidrio 

All tests now pass

when you get a chance, please look over this PR implementing the rows-last for multidimensional columns throughout MatNWB. It was quite a tricky thing to implement so I appreciate any feedback wherever thing could be more elegant.

Let me recap some of the main points I highlighted above:

-----------
1) In order to prevent breaking current users' scripts, single-dimensional columns should be defined with MATLAB's rows-first convention

2) For columns with more than one dimension, the desired row should come last. 

Here's a snippet illustrating these points

```matlab
% create NwbFile object with required fields
file= NwbFile( ...
    'session_start_time', '2021-01-01 00:00:00', ...
    'identifier', 'ident1', ...
    'session_description', 'DynamicTableTest' ...
    );

% create VectorData objects 
start_time= types.hdmf_common.VectorData( ...
    'description', 'start times column', ...
    'data', (1:6).' ...
);

stop_time = types.hdmf_common.VectorData( ...
    'description', 'stop times column', ...
    'data', (2:7).' ...
);

random = types.hdmf_common.VectorData( ...
    'description', 'random data column', ...
    'data', rand(4,3,6) ... % rows-last
);

ids = types.hdmf_common.ElementIdentifiers( ...
    'data', (0:5).' ...
);
% create expandable table
colnames = {'start_time', 'stop_time', 'randomvalues'};
file.intervals_trials = types.core.TimeIntervals( ...
    'description', 'test dynamic table', ...
    'colnames', colnames, ...
    'start_time', start_time, ...
    'stop_time', stop_time, ...
    'randomvalues', random, ...
    'id', ids ...
);  
% export file
nwbExport(file, 'DynamicTableTest.nwb');
```
------

3) usage of getRow and toTable methods. Given that 1)  the output of both these methods are MATLAB tables and 2) for these tables to be valid the height of all columns must be matched, the dimensions of multidimensional columns are swapped when these methods are called. The need for this swap is illustrated with the following snippet:

```matlab
start_times = types.hdmf_common.VectorData( ...
    'description', 'start_times', ...
    'data', 1:10 ...
);

stop_times = types.hdmf_common.VectorData( ...
    'description', 'stop_times', ...
    'data', 2:11 ...
);
randomvalues = types.hdmf_common.VectorData( ...
    'description', 'random values',...
    'data', rand(3,2,10) ...
);

multi_dim_table = types.core.TimeIntervals( ...
    'description','test table', ...
    'colnames', {'randomvalues', 'start_time', 'stop_time'}, ...
    'start_time', start_times, ...
    'stop_time', stop_times, ...
    'randomvalues', randomvalues, ...
     'id', types.hdmf_common.ElementIdentifiers('data', 0:9) ...
 );

>> multi_dim_table.toTable()

ans =

  10×4 table

    id     randomvalues     start_time    stop_time
    __    ______________    __________    _________

    0     [1×2×3 double]         1            2    
    1     [1×2×3 double]         2            3    
    2     [1×2×3 double]         3            4    
    3     [1×2×3 double]         4            5    
    4     [1×2×3 double]         5            6    
    5     [1×2×3 double]         6            7    
    6     [1×2×3 double]         7            8    
    7     [1×2×3 double]         8            9    
    8     [1×2×3 double]         9           10    
    9     [1×2×3 double]        10           11    

```

without the dimension swap, the 'randomvalues' column would have a height of 30 while all other columns have a height of 10. 

An open question, I would like to discuss is on how we want to direct users to define single-dimensional columns when DataPipes are involved. As you both have pointed out in the other PR, the mapping between MatNWB and HDF5 dimensions differs whether or not DataPipes are involved:

MatNWB            HDF5
======            =====
       -—VectorData—
[N,M]      <->    [M,N]
[N,M,O]    <->    [O,M,N]
[N,1]      <->    [N]
[1,N]      <->    [N]
impossible <->    [1,N]
       -—DataPipe--
[N,M]      <->    [M,N]
[N,M,O]    <->    [O,M,N]
[N,1]      <->    [1,N]
[1,N]      <->    [N,1]

That means that if one wants to end up with row vectors in the HDF5 file and subsequently in PyNWB, one should define single-dimensional columns with column vectors as follows:

```matlab
% create NwbFile object with required fields
file= NwbFile( ...
    'session_start_time', '2021-01-01 00:00:00', ...
    'identifier', 'ident1', ...
    'session_description', 'DataPipeTest' ...
    );

%create VectorData objects with DataPipe objects
start_time_exp = types.hdmf_common.VectorData( ...
    'description', 'start times column', ...
    'data', types.untyped.DataPipe( ...
        'data', 1:6, ...  # data must be numerical
        'maxSize', [1, Inf], ...
        'axis', 2 ...
    ) ...
);

stop_time_exp = types.hdmf_common.VectorData( ...
    'description', 'stop times column', ...
    'data', types.untyped.DataPipe( ...
        'data', 2:7, ...  #data must be numerical
        'maxSize', [1, Inf], ...
        'axis', 2 ...
    ) ...
);

random_exp = types.hdmf_common.VectorData( ...
    'description', 'random data column', ...
    'data', types.untyped.DataPipe( ...
        'data', rand(4, 3 , 6), ...  # rows last 
        'maxSize', [4, 3,  Inf], ...
        'axis', 3 ...
    ) ...
);

ids_exp = types.hdmf_common.ElementIdentifiers( ...
    'data', types.untyped.DataPipe( ...
        'data', 0:5, ...  # data must be numerical
        'maxSize', [1, Inf], ...
        'axis', 2 ...
    ) ...
);
% create expandable table
colnames = {'start_time', 'stop_time', 'randomvalues'};
file.intervals_trials = types.core.TimeIntervals( ...
    'description', 'test expdandable dynamic table', ...
    'colnames', colnames, ...
    'start_time', start_time_exp, ...
    'stop_time', stop_time_exp, ...
    'randomvalues', random_exp, ...
    'id', ids_exp ...
);  
% export file
nwbExport(file, 'dataPipeTest.nwb');
``` 
Screen Shot 2021-12-23 at 12 58 22 PM

Does this make sense to you? It might not be the most user-friendly thing to direct users to define single-dimensional columns as row-vectors in one context and as column-vectors in another, but will probably make things easiest for compatibility between MatNWB and PyNWB.

This is more of an issue for the hdmf tutorial rather than something that would break this PR. Let me know your thoughts.

File: +tests/+system/DynamicTableTest.m [hunk 1]
```diff
@@ -45,15 +28,15 @@ function addExpandableDynamicTable(~, file, start_array, stop_array, ...
             start_time_exp = types.hdmf_common.VectorData( ...
                 'description', 'start times', ...
                 'data', types.untyped.DataPipe( ...
-                    'data', start_array', ...
+                    'data', start_array, ...
                     'maxSize', [Inf, 1], ...
```
Comment (bendichter) [inline +tests/+system/DynamicTableTest.m:32]: I think `maxSize` should be flipped too

File: +tests/+system/DynamicTableTest.m [hunk 2]
```diff
@@ -45,15 +28,15 @@ function addExpandableDynamicTable(~, file, start_array, stop_array, ...
...
                     'axis', 1 ...
                 ) ...
             );
             stop_time_exp = types.hdmf_common.VectorData( ...
                 'description', 'stop times', ...
                 'data', types.untyped.DataPipe( ...
-                    'data', stop_array', ...
+                    'data', stop_array, ...
                     'maxSize', [Inf, 1], ...
```
Comment (bendichter) [inline +tests/+system/DynamicTableTest.m:40]: I think `maxSize` should be flipped too

File: +tests/+system/DynamicTableTest.m [hunk 3]
```diff
@@ -70,13 +53,13 @@ function addExpandableDynamicTable(~, file, start_array, stop_array, ...
...
-                    'axis', 1 ...
+                    'maxSize', [3, 2 , Inf], ...
+                    'axis', 3 ...
                 )...
             );
             ids_exp = types.hdmf_common.ElementIdentifiers( ...
                 'data', types.untyped.DataPipe( ...
-                    'data', id_array', ... 
+                    'data', id_array, ... 
                     'maxSize', [Inf, 1], ...
```
Comment (bendichter) [inline +tests/+system/DynamicTableTest.m:63]: I think `maxSize` should be flipped too

File: +tests/+system/DynamicTableTest.m [hunk 4]
```diff
@@ -105,11 +88,11 @@ function addContainerUnevenColumns(~, file)
                 'colnames', colnames, ...
                 'start_time',  types.hdmf_common.VectorData( ...
                     'description', 'start time column', ...
-                    'data', (1:200)' ...
+                    'data', (1:200).' ...
```
Comment (bendichter) [inline +tests/+system/DynamicTableTest.m:91]: I don't understand this. Why did you add a period? Shouldn't this be non-transposed?

File: +tests/+system/DynamicTableTest.m [hunk 5]
```diff
@@ -105,11 +88,11 @@ function addContainerUnevenColumns(~, file)
...
                     ), ...
                 'stop_time',  types.hdmf_common.VectorData( ...
                     'description', 'stop time column', ...
-                    'data', (2:101)' ...
+                    'data', (2:101).' ...
```
Comment (bendichter) [inline +tests/+system/DynamicTableTest.m:95]: I don't understand this. Why did you add a period? Shouldn't this be non-transposed?

**Comment by bendichter** [comment]
@cechava I don't understand this `(1:6).'` syntax you are using throughout. What are you trying to accomplish with that period?

File: +util/swapDims.m [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+%   array. Use when trailing singleton dimensions present in DATA array
+%   (e.g., 3x2x1).
+if nargin<2
+    dims = ndims(data);
+end
+dim_order = 1:dims;
+dim_order(1) = dims;
+dim_order(end) = 1;
+data = permute(data,dim_order);
+end
```
Comment (bendichter) [inline +util/swapDims.m:19]: This is not right. The data is entirely transposed. A [N,M,O,P] would become a [P,O,M,N], not a [P,M,O,N].

File: +types/+util/+dynamictable/getRow.m [hunk 7]
```diff
@@ -42,6 +42,18 @@
     end
 
     row{i} = select(DynamicTable, indexNames, ind);
+    if size(row{i},2)>1
+        % swap dimensions of non-row vectors. otherwise will result in
+        % invalid MATLAB table with uneven column height
+        row{i} = util.swapDims(row{i});
```
Comment (bendichter) [inline +types/+util/+dynamictable/getRow.m:48]: why do we need this?

File: +tests/+sanity/GenerationTest.m [hunk 8]
```diff
@@ -49,9 +49,11 @@ function dynamicTableMethodsTest(testCase)
             t = table(id(101:200), (101:200) .', (102:201) .',...
                 mat2cell(rand(500,1), repmat(5, 100, 1)), repmat({'TRUE'}, 100, 1),...
```
Comment (bendichter) [inline +tests/+sanity/GenerationTest.m:50]: could we change this so it's not generating hundreds of random variables?

**Comment by bendichter** [comment]
This creates incorrect tables!

image


@cechava , this PR has lots of mistakes. Let me know when you are available to discuss a path forward.

**Comment by bendichter** [comment]
> In order to prevent breaking current users' scripts, single-dimensional columns should be defined with MATLAB's rows-first convention

wait, what?

File: +tests/+system/DynamicTableTest.m (hunk 4 above)
Comment (cechava) [reply +tests/+system/DynamicTableTest.m:91]: I mindlessly added the period and copied the mistake at other points. The period is used for element-wise operations. In this context, it does nothing. I can clean that up

File: +util/swapDims.m (hunk 6 above)
Comment (cechava) [reply +util/swapDims.m:19]: You are right. I was thinking it was really the first and last dimensions that are relevant for these fixes, but I didn't think that, in the round trip, the middle dimensions would also be flipped. I can quickly fix this.

File: +types/+util/+dynamictable/getRow.m (hunk 7 above)
Comment (cechava) [reply +types/+util/+dynamictable/getRow.m:48]: will illustrate this in synchronous chat

File: +tests/+sanity/GenerationTest.m (hunk 8 above)
Comment (cechava) [reply +tests/+sanity/GenerationTest.m:50]: yes, can do

File: +tests/+system/DynamicTableTest.m (hunk 1 above)
Comment (cechava) [reply +tests/+system/DynamicTableTest.m:32]: correct, changing

**Comment by cechava** [comment]
> This creates incorrect tables!
> 
> image
> 
> @cechava, let me know when you are available to discuss a path forward.

can you share the snippet that you used to generate that file? I just want to make sure I'm not missing a bug that is not covered by the current tests

**Comment by cechava** [comment]
putting this on the shelf, for now
